op-challenger/fault/main.go

```diff
@@ -0,0 +1,179 @@
+package main
+
+import "fmt"
+
+// Important: need to know the full range b/c if we agree
+// with the middle & end, we don't want to do anything.
+// Also: what about different cases with agreeing/disagreeing?
+
+// Position is a golang wrapper around the dispute game Position type.
+// Depth refers to how many bisection steps have occurred.
+// IndexAtDepth refers to the path that the bisection has taken
+// where 1 = goes right & 0 = goes left.
+type Position struct {
+	Depth        int
+	IndexAtDepth int
+}
+
+// TraceIndex calculates the what the index of the claim value
+// would be inside the trace.
+func (p *Position) TraceIndex(maxDepth int) int {
+	if p.Depth == 0 {
+		return 1 << maxDepth
+	}
+	// TODO: I'm somehow doing an extra iteration (which is why I have to add 1 to the max depth).
+	// I think it might be working because we always go left at the start.
+	lo := 0
+	hi := 1 << (maxDepth + 1)
+	var mid int
+	path := p.IndexAtDepth
+	for i := p.Depth - 1; i >= 0; i-- {
+		mid = (lo + hi) / 2
+		mask := 1 << i
+		if path&mask == mask {
+			lo = mid
+		} else {
+			hi = mid
+		}
+	}
+	return (hi + lo) / 2
+}
+
+// IMPORTANT: This is not the same as in the contracts.
+func (p *Position) Move(right bool) {
+	p.Depth++
+	p.IndexAtDepth = (p.IndexAtDepth << 1) | boolToInt(right)
+}
+
+func boolToInt(b bool) int {
+	if b {
+		return 1
+	} else {
+		return 0
+	}
+}
+
+func (p *Position) Print(maxDepth int) {
+	fmt.Printf("Trace Position is %04b\tTrace Depth is: %d\tTrace Index is: %d\n", p.IndexAtDepth, p.Depth, p.TraceIndex(maxDepth))
+}
+
+// TraceProvider is a generic way to get a claim value at a specific
+// step in the trace. The trace value may be unavailable.
+// Get must not panic, but MustGet should panic if the value is not
+// available.
+type TraceProvider[T comparable] interface {
+	MustGet(i int) T
+	Get(i int) (*T, error)
+}
+
+// AlphabetProvider
+
+// ContractClaim is how a claim is represented in the contract
+type ContractClaim[T comparable] struct {
+	GetParent func() *ContractClaim[T]
+	Value     T
+	GoesLeft  bool
+	Position
+}
+
+func (c *ContractClaim[T]) IsRoot() bool {
+	return c.Depth == 0
+}
+
+type Response[T comparable] struct {
+	Value    T
+	GoesLeft bool
+}
+
+func counter[T comparable](theirs ContractClaim[T], trace TraceProvider[T]) *Response[T] {
+	index := theirs.Position.TraceIndex(3)
+	ours := trace.MustGet(index)
+	if ours == theirs.Value {
+		// go right
+		// walk until we know the right bound of this range. This would have been set by a claim going left or the root claim
+	} else {
+		// go left
+		// walk up the claims until we know the left bound. This will have been set by a node going right or the root claim.
```

Comment:
I think this logic is incorrect. We shouldn't need anything beyond the parent of the claim we're evaluating.  The bounds can be calculated purely based on the claim's position - you can see that happening in the solidity in `LibPosition`: https://github.com/ethereum-optimism/optimism/blob/84e8b55b521bc39eca8540e866b3e416f1615b96/packages/contracts-bedrock/contracts/dispute/lib/LibPosition.sol#L10

These comments sounds like they're trying to play a game more like "guess my number", where I can pick arbitrary numbers to guess and get told if it's too high or too low. Then I have to track what the upper and lower bounds are across multiple guesses to know what to guess next.

But actually each of my "guesses" is fixed - I have to pick the middle option every time. All I'm supplying is the value I have at that fixed index. So if the maximum possible trace length (defined by max depth) is 100 long, the root claim _must_ be from index 100.  The first counter claim _must_ be at index 50.  Then there are two choices depending on whether you attack or defend that first counter claim - either 25 or 75. If we responding to a claim at trace index 25, we must respond with a trace at either 12.5 or 37.5 (rounded down I think).

We can convert between `Position` and it's trace offset, and when we attack or defend we can calculate the `Position` of the new claim based on the parent claim. So we shouldn't ever need to walk back up the claims to determine bounds.

Note that the top level is a bit weird here because it's not actually a binary tree - the root claim only has one child (the hash half way along the padded trace), then the first counter claim is where the binary tree actually starts and each node from there down has two child positions. The other part that can cause confusion is that you can have multiple claims at the same position as long as they have different values. So my original intuition that we should build all the claims into a single model of the game state is a bad idea as it introduces all the complexity of these multiple claims at the same position.

Much better is to just think of each individual claim with a small tree around it.  We'd just have:

```
                                       Parent
                                          |
                                        Claim
                                       /     \
                                     /        \
                                    /          \
                               Attack?      Defend?
``` 
The "Attack?" and "Defend?" are the positions we can create a counter claim at if we need to. So we have two inputs: `Parent(Position, Value), Claim(Position, Value)` and the output is either "do nothing", `Attack(Value)` or `Defend(Value)`, and we need to iterate over every claim in the game and make that decision (then drop any actions we want to make that already exist in the game).  So something like:

```go

type Claim struct {
	Value common.Hash
	Position // Reusing Position from PR code.
}

type Response struct {
	Attack bool
	Value  common.Hash
}

// Decide determines the response to claim.
// If the returned Response is nil, no counter claim is required.
func Decide(parent Claim, claim Claim) *Response {
	parentCorrect := agreeWithClaim(parent)
	claimCorrect := agreeWithClaim(claim)
	if parentCorrect && claimCorrect {
		// We agree with the parent, but the claim is disagreeing with it.
		// Since we agree with the claim, the difference must be to the right of the claim
		return defend(claim)
	} else if parentCorrect && !claimCorrect {
		// We agree with the parent, but the claim disagrees with it.
		// Since we disagree with the claim, the difference must be to the left of the claim
		return attack(claim)
	} else if !parentCorrect && claimCorrect {
		// Do nothing, we disagree with the parent, but this claim has correctly countered it
		return doNothing()
	} else if !parentCorrect && !claimCorrect {
		// We disagree with the parent so want to counter it (which the claim is doing)
		// but we also disagree with the claim so there must be a difference to the left of claim
		// Note that we will create the correct counter-claim for parent when it is evaluated, no need to do it here
		return attack(claim)
	}
	panic("I think I handled all the options...")
}

func doNothing() *Response {
	return nil
}

func attack(claim Claim) *Response {
    // From LibPosition:
	// Left = { depth: position.depth + 1, indexAtDepth: position.indexAtDepth * 2 }
	counterPosition := Position{
		Depth:        claim.Position.Depth + 1,
		IndexAtDepth: claim.Position.IndexAtDepth * 2,
	}
	value := traceAtPosition(counterPosition)
	return &Response{Attack: true, Value: value}
}

func defend(claim Claim) *Response {
    // From LibPosition:
	// Right = { depth: position.depth + 1, indexAtDepth: position.indexAtDepth * 2 + 1 }
	counterPosition := Position{
		Depth:        claim.Position.Depth + 1,
		IndexAtDepth: claim.Position.IndexAtDepth*2 + 1,
	}
	value := traceAtPosition(counterPosition)
	return &Response{Attack: false, Value: value}
}

func agreeWithClaim(claim Claim) bool {
	ourValue := traceAtPosition(claim.Position)
	return ourValue == claim.Value
}

// traceAtPosition returns the common.Hash from the trace provider that we believe should be the value at p
func traceAtPosition(p Position) common.Hash {
	index := positionToTraceIndex(p)
    // TODO: Error handling and actually making the traceProvider available here...
	return traceProvider.Get(index)
}

// positionToTraceIndex calculates the index in the trace to use at a specific position
func positionToTraceIndex(p Position) uint64 {
   // I think this is actually the PR's `Position.TraceIndex`
}
```

The root claim would need a little special handling because it doesn't have a parent in the game - we have an honest proposer assumption currently so we always agree with the root claim's parent. When we remove the honest proposer assumption we'd have to check if we agree with the output root this game is challenging to know if we agree with the root claim's parent. We will also need special handling around the max depth where we need to switch to calling `step()` instead of `attack` or `defend`, but we're still evaluating just the claim and it's parent to decide if we need to `step` - just that we don't have a choice between attack/left and defend/right when we get to max depth.

@clabby Do I have that right?

Reply:
> These comments sounds like they're trying to play a game more like "guess my number", where I can pick arbitrary numbers to guess and get told if it's too high or too low. Then I have to track what the upper and lower bounds are across multiple guesses to know what to guess next.

> But actually each of my "guesses" is fixed - I have to pick the middle option every time. All I'm supplying is the value I have at that fixed index. So if the maximum possible trace length (defined by max depth) is 100 long, the root claim must be from index 100. The first counter claim must be at index 50. Then there are two choices depending on whether you attack or defend that first counter claim - either 25 or 75. If we responding to a claim at trace index 25, we must respond with a trace at either 12.5 or 37.5 (rounded down I think).

If you keep track of the position of a node, you're correct that based on agreeing or disagreeing (defend / attack) you can find the position & index of the next value that you need to claim. The position logic is doing this averaging to find the specific trace value. I guess I was trying to keep the position logic out & then by dealing with trace index values you'd have to do this bisection step (you need to know the depth / side bound) to determine how large of a step to take.